pkg/gadgets/socket-collector/gadget.go

```diff
@@ -72,71 +75,82 @@ func (t *Trace) Operation(trace *gadgetv1alpha1.Trace,
 	params map[string]string) {
 
 	if trace.ObjectMeta.Namespace != gadgets.TRACE_DEFAULT_NAMESPACE {
-		gadgets.SetStatusError(trace, fmt.Sprintf("This gadget only accepts operations on traces in the %s namespace", gadgets.TRACE_DEFAULT_NAMESPACE))
+		gadgets.SetStatusError(trace, fmt.Sprintf("This gadget only accepts operations on traces in the %s namespace",
+			gadgets.TRACE_DEFAULT_NAMESPACE))
 		return
 	}
 
 	switch operation {
 	case "start":
-		var pid uint32 = 0
-
 		if trace.Spec.Filter == nil {
 			gadgets.SetStatusError(trace, "Missing filter")
 			return
 		}
-		if trace.Spec.Filter.Namespace == "" {
-			gadgets.SetStatusError(trace, "Invalid filter: missing namespace")
-			return
-		}
-		if trace.Spec.Filter.Podname == "" {
-			gadgets.SetStatusError(trace, "TODO: Filtering only by namespace is not currently supported")
-			return
-		}
-		if len(trace.Spec.Filter.Labels) != 0 {
-			gadgets.SetStatusError(trace, "TODO: Filtering by labels is not currently supported")
-			return
-		}
 		if trace.Spec.Filter.ContainerName != "" {
-			log.Warningf("Gadget %s: Container name filter is not applicable in this gadget, ignoring it!", trace.Spec.Gadget)
+			log.Warningf("Gadget %s: Container name filter is not applicable in this gadget, ignoring it!",
+				trace.Spec.Gadget)
 		}
 
-		pidsInPod := t.resolver.LookupPIDByPod(trace.Spec.Filter.Namespace, trace.Spec.Filter.Podname)
-
-		// All containers inside a pod share the same network namespace
-		// thus we can just take the first valid pid we found
-		for _, pid = range pidsInPod {
-			if pid != 0 {
-				break
-			}
+		selector := gadgets.ContainerSelectorFromContainerFilter(trace.Spec.Filter)
+		filteredContainers := t.resolver.GetContainersBySelector(selector)
+		if len(filteredContainers) == 0 {
+			gadgets.SetStatusError(trace, fmt.Sprintf("No container matches the requested filter: %+v",
+				*trace.Spec.Filter))
+			return
 		}
```

Comment:
I'm wondering how we are handling this case in the process-collector @alban.

Reply:
The process-collector and the socket-collector works differently: they both use BPF Iterators that are specific to namespaces: to a pid namespace and to a network namespace respectively. Since network namespaces are not hierarchical, the gadget needs to enter the netns to get the data, and it needs to do that as many times as there are pods. Pid namespaces are hierarchical and the gadget pod runs with hostPID=true, so the BTF iterator is executed only one time and the filtering is done with the "containers" BPF map prepared by the gadget tracer manager. So the process-collector gadget code does not need to do anything with GetContainersBySelector etc., it is just automatic :)

Reply:
So if there is not any matching container it prints an empty list. I'm wondering if in that case we should also say that there is not any matching container :thinking: . For the process collector it's a actually not a big issue, if there is an empty list then there is not any matching container because at least there is one process per container. For the socket collector it's more complicated, how can the user distinguish between the case where there's not a matching container and the case there is not any open socket in the matched containers? 

In summary, I think that we should uniform the behavior of all collector when there's not a matching container even if their internal implementation is very different.

Reply:
I agree we should uniform them and given that, as you already said, an empty list on socket-collector could be misunderstood, I would suggest to modify the process-collector to also print an error. Do you all agree?

Reply:
> I would suggest to modify the process-collector to also print an error. Do you all agree?

I am still hesitating about this one: if you add a check on whether the pod exists, the check is not going to be atomically done at the same time as the execution of the BPF iterator, so we could get inconsistent results.

Could that be done client-side: if kubectl-gadget receives `[]`, then it could do another call to the API server to check if the pod exists, and if not, add a message? But if the user requested `--json`, then no message?

Reply:
I don't have a strong opinion here - adding a check client-side is also inconsistent... I'll defer to your judgment.

Reply:
I think it's something we can track later on.

Reply:
I created issue https://github.com/kinvolk/inspektor-gadget/issues/253 to track this.